src/tactic/suggest.lean

```diff
@@ -109,40 +111,40 @@ do env ← get_env,
 
 /--
 Apply the lemma `e`, then attempt to close all goals using
-`solve_by_elim { discharger := discharger }`, failing if `close_goals = tt`
+`solve_by_elim { ..opt }`, failing if `close_goals = tt`
 and there are any goals remaining.
 -/
 -- Implementation note: as this is used by both `library_search` and `suggest`,
 -- we first run `solve_by_elim` separately on a subset of the goals,
 -- whether or not `close_goals` is set,
 -- and then if `close_goals = tt`, require that `solve_by_elim { all_goals := tt }` succeeds
 -- on the remaining goals.
-meta def apply_and_solve (close_goals : bool) (discharger : tactic unit) (e : expr) : tactic unit :=
+meta def apply_and_solve (close_goals : bool) (opt : opt := { }) (e : expr) : tactic unit :=
 apply e >>
 -- Phase 1
 -- Run `solve_by_elim` on each "safe" goal separately, not worrying about failures.
 -- (We only attempt the "safe" goals in this way in Phase 1. In Phase 2 we will do
 -- backtracking search across all goals, allowing us to guess solutions that involve data, or
 -- unify metavariables, but only as long as we can finish all goals.)
-try (any_goals (independent_goal >> solve_by_elim { discharger := discharger })) >>
+try (any_goals (independent_goal >> solve_by_elim { ..opt })) >>
 -- Phase 2
 (done <|>
   -- If there were any goals that we did not attempt solving in the first phase
   -- (because they weren't propositional, or contained a metavariable)
   -- as a second phase we attempt to solve all remaining goals at once
   -- (with backtracking across goals).
   any_goals (success_if_fail independent_goal) >>
-  solve_by_elim { backtrack_all_goals := tt, discharger := discharger } <|>
+  solve_by_elim { ..opt } <|>
```

Comment:
@u6338192, when you removed the `backtrack_all_goals` here, you broke `suggest`. Did you run the `test/suggest.lean` file locally? It would have shown you this problem.

Reply:
I set `backtrack_all_goals ;= tt` in the interactive functions for `library_search` and `suggest`, and then pass that through all the functions using `..opt`.

```lean
meta def library_search (hs : parse simp_arg_list) (attr_names : parse with_ident_list)
  (opt : opt := { }) : tactic unit :=
do asms ← mk_assumption_set ff hs attr_names,
   tactic.library_search
     {backtrack_all_goals := tt,
      lemma_thunks := return asms,
      ..opt} >>=
   if is_trace_enabled_for `silence_library_search then
     (λ _, skip)
   else
     trace
```

I was only looking at test/library_search/basic.lean. I didn't look at test/suggest.lean.

Reply:
The problem is that test/suggest.lean calls `suggest_scripts` directly.

One possible fix is to replace `suggest_scripts` with `suggest_scripts none {backtrack_all_goals := tt}` in all the failing tests.

Another option is to replace the line `do L ← suggest limit opt,` in the `suggest_scripts` functions with `do L ← suggest limit {backtrack_all_goals := tt, ..opt},`. As well as remove `backtrack_all_goals := tt, ` from the `library_search` and `suggest` interactive functions. This is probably better code anyway, because there's less similar code. 

The last option is to just put `{backtrack_all_goals := tt, ..opt},` back into `apply_and_solve`. I don't know if this is good coding practice though, because `apply_and_solve` is much further downstream than `suggest_scripts`.

I can do the coding, but I don't want to commit anything from my local machine. I'm not yet quite sure how git works and I'm afraid of test/library_search/basic.lean. Let me know what you think.